### PR TITLE
Force regression merge

### PR DESCRIPTION
**Pull request by viviannna**
Added force regression functionality. Works for NinaproDB2 and NinaproDB3 using exercise 3 data. 

Main Changes: 
- Now keeps X_, Y_, and label_ instead of just X_ and Y_. Any time there is a Y_ (ex: Y_test) added a corresponding label_ (ex: label_test). This allows force_regression to use labels as stratify. 
- Updated the train_test_split function to take in variable type for Y and to always return train and test for X, y, and label. 
- Added make_gestures_sequential to ensure no gaps between exercise labels. Prevents out of bounds issues in train_test_split.py.
- Modified getExtrema function to be based on windows similarly to the train_test_split.py instead of the unfold=False logic. 
- Generalized the pytorch training loop. get_metrics() returns list of metrics to use which are iterated through now (instead of manually declaring metrics). Works for MLP and resnet18. 
- When using NinaproDB3 for force regression, subject 10 of the study is deleted as they are missing the majority of force exercises.

### CODE REVIEW

**Review by emgbench — COMMENTED** [review]
Leaving some comments.

File: utils_NinaproDB2.py [hunk 1]
```diff
@@ -210,59 +239,153 @@ def target_normalize (data, target_min, target_max, restim):
...
+            d = decrements[exercise]
+    
+            balanced_restim[x][0][0] = value - d
+
+    return balanced_restim
+
 def getExtrema (n, p, exercise):
-    emg = getEMG((n, exercise), unfold=False)
-    labels = getLabels((n, exercise), unfold=False)
+    """Returns the min max of the electrode per gesture for a proportion of its windows.
```
Comment (emgbench) [inline utils_NinaproDB2.py:325]: good job fixing getExtrema. Could you do it for all the utils?

**Review by emgbench — COMMENTED** [review]
Really good work. I liked the refactoring for metrics and the simple looking fix for getExtrema. I had a few questions I wrote as comments.

File: CNN_EMG.py [hunk 2]
```diff
@@ -197,12 +197,23 @@ def list_of_ints(arg):
...
     project_name = 'emg_benchmarking_ninapro-db3'
     exercises = True
     if args.leave_one_session_out:
         raise ValueError("leave-one-session-out not implemented for ninapro-db3; only one session exists")
+    
+    if args.force_regression:
+        print("NOTE: Subject 10 is missing gesture data for exercise 3. Because of this, subject is ignored for regression and subject 11 is relabeled as subject 10.")
+        MISSING_SUBJECT = 10 # subject 10 missing most force data
+
+    assert not (args.force_regression and args.leftout_subject == 11), "Subject 11 is relabeled as participant 10 in regression "
```
[the remark this answers is not in the document]
Comment (viviannna) [reply CNN_EMG.py:211]: Updated to instead flag utils.leftout_subject == 10. User's arguments/config will line up with the paper's labeling. After taking inputs, subject 10 is then deleted internally such that subject 11 becomes subject 10. Any future references to subjects in the program assumes this change.

File: cross_validation_utilities/train_test_split.py [hunk 3]
```diff
@@ -44,29 +83,40 @@ def train_test_split(
...
         y_train = np.concatenate(y_train)
         y_test = np.concatenate(y_test)
+        label_train = np.concatenate(label_train)
+        label_test = np.concatenate(label_test)
         
         # if Y_train_set is one-hot-encoded, turn results into one-hot-encoded
-        if Y_train_set.shape[1] != 1:
-            y_train = np.eye(Y_train_set.shape[1])[y_train]
-            y_test = np.eye(Y_train_set.shape[1])[y_test]
+        if not force_regression and Y_train_set_og.shape[1] != 1:
```
[the remark this answers is not in the document]
Comment (viviannna) [reply cross_validation_utilities/train_test_split.py:109]: Defined NUM_GESTURES based on Y_train_set.shape[1] and use it instead of np.unique so sizes are consistent regardless of missing data.

File: ml_metrics_utils.py
[the remark this answers is not in the document]
Comment (viviannna) [reply ml_metrics_utils.py:1]: testing_metrics is a list of torch metrics returned by get_metrics in CNN_EMG.py. I added it in as a new argument that way you only ever have to modify get_metrics.

**Comment by jehanyang** [comment]
This fixes target normalization for all datasets and implements force-based regression for two Ninapro datasets.
